fuse_models/src/odometry_2d_publisher.cpp

```diff
@@ -245,63 +248,137 @@ void Odometry2DPublisher::tfPublishTimerCallback(const ros::TimerEvent& event)
+
+      // TODO(efernandez) for now we set the top right and bottom left corners to zero, but we could cache them in
+      // another attribute when we retrieve the covariance from the ceres problem
+      covariance.topRightCorner<3, 3>().setZero();
+      covariance.bottomLeftCorner<3, 3>().setZero();
+
+      const Eigen::Matrix<double, 6, 6> J = jacobian.topLeftCorner<6, 6>();
+      covariance = J * covariance * J.transpose();
+      covariance.noalias() += dt * params_.process_noise_covariance.topLeftCorner<6, 6>();
+
+      odom_output_.pose.covariance[0] = covariance(0, 0);
+      odom_output_.pose.covariance[1] = covariance(0, 1);
+      odom_output_.pose.covariance[5] = covariance(0, 2);
+      odom_output_.pose.covariance[6] = covariance(1, 0);
+      odom_output_.pose.covariance[7] = covariance(1, 1);
+      odom_output_.pose.covariance[11] = covariance(1, 2);
+      odom_output_.pose.covariance[30] = covariance(2, 0);
+      odom_output_.pose.covariance[31] = covariance(2, 1);
+      odom_output_.pose.covariance[35] = covariance(2, 2);
+
+      odom_output_.twist.covariance[0] = covariance(3, 3);
+      odom_output_.twist.covariance[1] = covariance(3, 4);
+      odom_output_.twist.covariance[5] = covariance(3, 5);
+      odom_output_.twist.covariance[6] = covariance(4, 3);
+      odom_output_.twist.covariance[7] = covariance(4, 4);
+      odom_output_.twist.covariance[11] = covariance(4, 5);
+      odom_output_.twist.covariance[30] = covariance(5, 3);
+      odom_output_.twist.covariance[31] = covariance(5, 4);
+      odom_output_.twist.covariance[35] = covariance(5, 5);
+    }
   }
 
-  trans.transform.translation.x = pose.getX();
-  trans.transform.translation.y = pose.getY();
-  trans.transform.translation.z = odom_output_.pose.pose.position.z;
-  trans.transform.rotation = tf2::toMsg(pose.getRotation());
+  odom_pub_.publish(odom_output_);
 
-  if (params_.world_frame_id == params_.map_frame_id)
+  if (params_.publish_tf)
   {
-    try
+    geometry_msgs::TransformStamped trans;
+    trans.header = odom_output_.header;
+    trans.child_frame_id = odom_output_.child_frame_id;
+    trans.transform.translation.x = pose.getX();
+    trans.transform.translation.y = pose.getY();
+    trans.transform.translation.z = odom_output_.pose.pose.position.z;
+    trans.transform.rotation = tf2::toMsg(pose.getRotation());
+
+    if (params_.world_frame_id == params_.map_frame_id)
     {
-      auto base_to_odom = tf_buffer_->lookupTransform(
-        params_.base_link_frame_id,
-        params_.odom_frame_id,
-        latest_stamp_,
-        params_.tf_timeout);
-
-      geometry_msgs::TransformStamped map_to_odom;
-      tf2::doTransform(base_to_odom, map_to_odom, trans);
-      map_to_odom.child_frame_id = params_.odom_frame_id;
-      trans = map_to_odom;
+      try
+      {
+        auto base_to_odom = tf_buffer_->lookupTransform(
+          params_.base_link_frame_id,
+          params_.odom_frame_id,
+          trans.header.stamp,
```

Comment:
@svwilliams @ayrton04 This was a bug! Before the time stamp passed to `lookupTransform` was `latest_stamp_`, but if `predict_to_current_time` is `true`, this should be `event.current_real` instead. Using `trans.header.stamp` we get the correct stamp in both cases.

Because of this, with a robot turning in place the pose estimation overshoots. This is easy to see looking at a LIDAR scan alignment wrt the map. After this fix, the scans aligned perfectly. :smiley: 

Let me know if you want this on another PR, only with this change, as it looks like a big issue.